src/expr/core/src/expr/external/iceberg.rs

```diff
@@ -0,0 +1,399 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt::{Debug, Formatter};
+use std::sync::Arc;
+
+use icelake::types::{
+    Bucket as BucketTransform, Day as DayTransform, Hour as HourTransform, Month as MonthTransform,
+    TransformFunction, Truncate as TruncateTransform, Year as YearTransform,
+};
+use risingwave_common::array::{ArrayRef, DataChunk};
+use risingwave_common::ensure;
+use risingwave_common::row::OwnedRow;
+use risingwave_common::types::{DataType, Datum};
+use risingwave_expr::expr::{get_children_and_return_type_for_func_call, BoxedExpression, Build};
+use risingwave_expr::Result;
+use risingwave_pb::expr::ExprNode;
+
+/// This module contains the iceberg expression for computing the partition value.
+/// spec ref: <https://iceberg.apache.org/spec/#partition-transforms>
+
+pub struct Bucket {
+    child: BoxedExpression,
+    n: i32,
+    transform: BucketTransform,
+}
+
+impl Debug for Bucket {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Iceberg_Bucket({})", self.n)
+    }
+}
+
+#[async_trait::async_trait]
+impl risingwave_expr::expr::Expression for Bucket {
+    fn return_type(&self) -> DataType {
+        DataType::Int32
+    }
+
+    async fn eval(&self, data_chunk: &DataChunk) -> Result<ArrayRef> {
+        // Get the child array
+        let array = self.child.eval(data_chunk).await?;
+        // Convert to arrow array
+        let arrow_array = array.as_ref().try_into().unwrap();
+        // Transform
+        let res_array = self.transform.transform(arrow_array).unwrap();
+        // Convert back to array ref and return it
+        Ok(Arc::new((&res_array).try_into().unwrap()))
+    }
+
+    async fn eval_row(&self, _row: &OwnedRow) -> Result<Datum> {
+        unimplemented!()
+    }
+}
+
+impl Build for Bucket {
+    fn build(
+        prost: &ExprNode,
+        build_child: impl Fn(&ExprNode) -> Result<BoxedExpression>,
+    ) -> Result<Self> {
```

Comment:
The build function can be written as:
```rust
#[build_function("iceberg_bucket(any, int4) -> int4")]
fn build(return_type: DataType, children: Vec<BoxedExpression>) -> Result<BoxedExpression> {
    ...
}
```

After that, this file can be moved into `expr_impl` crate.

Reply:
Nice suggestion! I have refactor using the build function.